stablehlo/reference/InterpreterApi.cpp

```diff
@@ -0,0 +1,126 @@
+limitations under the License.
+==============================================================================*/
+#include "stablehlo/reference/InterpreterApi.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "stablehlo/reference/Errors.h"
+#include "stablehlo/reference/InterpreterOps.h"
+#include "stablehlo/reference/NumPy.h"
+#include "stablehlo/reference/Ops.h"
+
+namespace mlir {
+namespace stablehlo {
+
+llvm::Error InterpreterFallback::operator()(Operation &op, Process *process,
+                                            Scope &scope) {
+  llvm::StringRef funcName = currentFunction.getSymName();
+
+  if (auto probeOp = dyn_cast<stablehlo::interpreter::ProbeOp>(op)) {
+    auto input =
+        stablehlo::InterpreterValue(scope.findTensor(probeOp.getOperand()));
+    auto status = stablehlo::interpreter::evalProbeOp(
+        input, probeOp.getProbeId(), config->probeInstrumentationDir,
+        instrumentedTensors);
+    scope.add(probeOp.getResult(), input);
+    return wrapStatus(std::move(status), funcName, "interpreter.probe");
+  }
+
+  if (auto runParallelOp =
+          dyn_cast<stablehlo::interpreter::RunParallelOp>(op)) {
+    auto runtimeOperands = scope.find(runParallelOp.getInputs());
+    std::queue<StringAttr> infeed;
+    if (auto infeedAttr = runParallelOp.getInfeed())
+      for (auto &value : infeedAttr->getValue())
+        infeed.push(value.cast<FlatSymbolRefAttr>().getAttr());
+
+    SmallVector<SmallVector<StringAttr>> programs(
+        runParallelOp.getPrograms().size());
+    for (auto [i, replica] : llvm::enumerate(runParallelOp.getPrograms()))
+      for (auto &program : replica.cast<ArrayAttr>())
+        programs[i].push_back(program.cast<FlatSymbolRefAttr>().getAttr());
+
+    SymbolTable symbolTable{op.getParentOfType<ModuleOp>()};
+    auto results = stablehlo::interpreter::evalRunParallelOp(
+        runtimeOperands, infeed, programs, symbolTable);
+    scope.add(runParallelOp.getResults(), results);
+    return wrapStatus(llvm::Error::success(), funcName,
+                      "interpreter.run_parallel");
+  }
+
+  return handleOp(op, process, scope);
+}
+
+llvm::Error InterpreterFallback::handleOp(Operation &op, Process *process,
+                                          Scope &scope) {
+  return stablehlo::invalidArgument("Unsupported op: %s",
+                                    debugString(op).c_str());
+}
+
+llvm::ErrorOr<SmallVector<InterpreterValue>> runInterpreter(
+    ModuleOp module, ArrayRef<InterpreterValue> inputs,
+    const InterpreterConfiguration &config) {
+  auto numFuncs = 0;
+  bool hasMain = false;
+
+  assert(!config.mainFunction.empty() && "Main function name cannot be empty");
+  module.walk([&](func::FuncOp funcOp) {
```

Comment:
I think to preserve the current behavior (i.e. for existing LIT tests), the semantics of the function need to be if there is only 1 function in the module, use that as the main function. Otherwise, use the provided `mainFunction` string to break ties when there are more than 1 function. Slightly tweaked proposed change with this.